### ISSUE TITLE
Rename content-addressed derivations's "realisation" to "trace"

### ISSUE THREAD

**Issue by Ericson2314** [NixOS/nix#11895]
"realisation" is a bad term because we already use it for other things (building, including input-addressed derivations) and because it is vague.

The term in https://www.microsoft.com/en-us/research/uploads/prod/2018/03/build-systems.pdf and https://cloud.ins.jku.at/index.php/s/txgoHps6FyNpiDk "trace"; we should use that instead, in documentation and code.

**Comment by mschwaig** [comment]
closes https://github.com/NixOS/nix/issues/7254, which has some alternative suggestions

**Comment by edolstra** [comment]
"Realisation" is not a great term, but "trace" is no better, IMHO. It really tells me nothing about what it is.

**Comment by roberth** [comment]
Ambiguous with `--show-trace`

**Comment by Ericson2314** [comment]
Fair point. Well then there we go: build trace and eval trace!

**Comment by mschwaig** [comment]
Another alternative name that is a bit more distrinct would be kvtrace, with the k and v standing for key and value.
That highlights how each kvtrace represents an entry in a map from build inputs to build outputs.

**Comment by Ericson2314** [comment]
Maybe the whole thing should be called "build trace" or "build trace map"? and an individual one is a "build trace kv pair"?

**Comment by roberth** [comment]
We might actually be talking about two representations for information about builds.
- the occurrence of a build, including output content for the whole derivation
- a record that maps (drv, outname) -> output content

Excuse my ignorance about undocumented features, but I believe the prior is a realisation, whereas the latter is the more useful and accurate concept for reasoning about build inputs, and perhaps this is the mapping that deserves to be called a build trace?
These concepts convey largely the same information, but they give rise to a slightly different data model (or different _part_ of a data model).

- **realisation**: drv -> (map outname contenthash)
- **build trace**: (drv, output) -> contenthash

To make the distinction a bit more concrete, consider a derivation graph where one of the derivations has an unused `testresults` output, containing a timestamp. By comparing _realisations_, you would falsely flag a frankenbuild, whereas a set of traces would accurately report "no frankenbuild", because none of the impure outputs were used.

**Comment by roberth** [comment]
Relevant quote from the notes:

@mschwaig: using a realisation both for the process and the data may be ambiguous. Makes it hard to talk about the process vs the data.

**Comment by Ericson2314** [comment]
@roberth Good point, I too would prefer to query the map. There is little-to-no valid use-case to upload a partial output map to the cache; best just upload the whole thing or none of it.

**Comment by Ericson2314** [comment]
> Famous last words ;)
> Your assumption is that the number of outputs is small.

Yes but, *knocks on wood* I feel OK about this one.

> Have you considered derivations with large numbers of outputs for dynamic-derivations evaluating whole package sets?

I think the common idiom will be directory trees that have symlinks to other derivations for things that produce an statistically known number / shape of derivations.

In that world, I think statically declared multiple outputs might seen as the analogue to an "unboxed tuple", whereas symlinks to other store objects would be "regular boxed heap object".

> The binary cache protocol is relevant, but it doesn't necessarily have to decide how we name concepts.

True. Basically my thinking is: 

1. in order to not let CA drag on forever, I am trying to prioritize cache hygiene over things like client efficiency, see https://discourse.nixos.org/t/content-addressed-nix-call-for-testers/12881/228
2. We can still forget about the binary cache and instead think about things like signing, should we sign each output's mapping separately? But what is that for? If you actually did the build, then you should be able to sign off all the inputs at once.

> - **realisation**: drv -> (map outname contenthash)
>
> - **build trace**: (drv, output) -> contenthash

Actually the thing we have today is I think the second. On the other-hand I misread and thought I was agreeing with you not disagreeing :).
